### PR TITLE
Work in progress/nx integration

### PR DESCRIPTION
**Pull request by danielstevenberger**
Setup Nx for the Golden Sample App staying very close to their standardized workspace.

Notes:

1. Jest testing example without using the Angular Test Bed. 

- This was my first time using Jest, so I still need to learn best practices. My first thoughts are that it's significantly faster and the Visual Studio Extension makes writing and testing enjoyable. I believe a switch to this would allow all developers to have a much better experience writing unit tests that focus on the logic of the components and services.

2. Cypress was initialized in the project, but I have yet to touch it.

3. Nx cloud is currently disabled.

4. This is a big refactor. If it's decide to go through with this, all code should be frozen in the meantime. 

Feel free to ask me any questions or setup a meeting on my calendar anytime after 9am EST.

### CODE REVIEW

File: angular.json
Comment (georgejolema) [inline angular.json:185]: why this prefix?

File: README2.md
Comment (georgejolema) [inline README2.md:1]: why did you create this file? can this replace README?

File: yarn.lock
Comment (georgejolema) [inline yarn.lock]: remove this file

File: package.json
Comment (georgejolema) [inline package.json:6]: restore xi18n and docker scripts

File: package.json
Comment (georgejolema) [inline package.json:75]: this is still necessary for our translations

File: README2.md
Comment (yenobi) [reply README2.md:1]: It looks like it is here only to save the data from the original file, they should be merged, right? @danielstevenberger

**Comment by yenobi** [comment]
I would vote to remove everything related to Cypress, because QA guild is still discussing the recommended way of e2e.

**Comment by danielstevenberger** [comment]
Updates:

1. Previous PR Comments:
I believe I have resolved all of the comments. Please let me know if you have any issues with the translations or docker containers. They seem to be building correctly, but I was having trouble starting them on this branch and master as well. I'm not sure if it's an Nx or a preexisting issue. 

2. Angular Update
I've updated the project to Angular 13 and have everything working correctly on npm run start.
This was to test Marco's concern with Jest. There seems to be no issues.
There were some updates to eslint. Throughout the project there are some lint errors now, but I will leave this until the current state of this branch is approved.

3. Unit Test and e2e
I have resolved all of the Unit Test errors, but some places need more coverage. I've gotten some confirmation on Jest, but do not want to continue until there's a definite answer. I will also being leaving cypress as is since I've heard both feedback for removing and keeping it as well.

At this point please feel free to pull the branch and run npm i. You'll likely want to delete node_modules and the package-lock.json if you do this.

File: nx.json
Comment (jordydejong) [inline nx.json:1]: Do we want add tags, and impose constraints? https://nx.dev/structure/monorepo-tags
Currently it seems I am allowed to import from feature X in feature Y.

File: tsconfig.base.json
Comment (danielstevenberger) [reply tsconfig.base.json:18]: Is there a specific naming convention that you'd like me to follow?

File: .vscode/extensions.json
Comment (danielstevenberger) [reply .vscode/extensions.json:1]: I can add this to .gitignore

File: tsconfig.base.json
Comment (jordydejong) [reply tsconfig.base.json:18]: @santam85 has there ever been a decision made on this? the naming of journeys in an NX repo, in regards to the best practice of using "feature libraries"

**Comment by tcorral** [comment]
Information: Today's QA Guild has sent a message informing that Backbase is adopting Playwright as the tool for e2e tests

File: .vscode/extensions.json
Comment (tcorral) [reply .vscode/extensions.json:1]: @danielstevenberger , please, delete the folder (copy it on your local machine) in repo and then if you update .gitignore you will be able to keep using it on your local development environment without impacting others.

**Comment by danielstevenberger** [comment]
Introduce Playwright as option for E2E testing

From what I've read so far, Nx does not support or come with @playwright/test, but they can coexist. 

Here's a related issue I commented on.
https://github.com/nrwl/nx/issues/3390

Here's a plugin someone made to run with nx (This does not seem like the best option)
https://github.com/playwright-community/jest-playwright

**Comment by danielstevenberger** [comment]
> Introduce Playwright as option for E2E testing
> 
> From what I've read so far, Nx does not support or come with @playwright/test, but they can coexist.
> 
> Here's a related issue I commented on. [nrwl/nx#3390](https://github.com/nrwl/nx/issues/3390)
> 
> Here's a plugin someone made to run with nx (This does not seem like the best option) https://github.com/playwright-community/jest-playwright

File: package.json
Comment (georgejolema) [inline package.json:19]: why did you update to angular 13? if this is a sample for all the developers, we should keep the state of this the same as developers are doing, so that they can base this as their source of truth. 

Maybe we can keep an unstable branch or something with the latest angular for research and testing.

Comment (santam85) [reply package.json:19]: Daniel did it to test if there was an issue with jest for testing. If it's working ok I don't think we need to necessarily revert it, we want to showcase "future state" as well

Comment (georgejolema) [reply package.json:19]: I am fine to do this testing for "future state" but not for something that is a reference to other developers. Here we show how to handle journeys, authentication, entitlements, deploying, communication, data handling, core libraries implementations, building configurations, best practices and now monorepo setup. All of this is based on angular, if you use a future version, that would also force us to upgrade the peer dependencies. And therefore this will no longer reflect what we want developers to do in the capabilities. But fortunately we can make a branch for future state and experiment there. This is something we all agree in web-foundation during our last meeting.

Comment (tcorral) [reply package.json:19]: In Master branch, @santam85, I would like to keep the current version but anyways this doesn't stop us of having an experimental branch on which we can start working updating Angular as well as any other things we would like to add to master sooner or later. We do link the code inside our documentation in community and if we do changes like this one it might end with confusing code versus documentation.

Comment (santam85) [reply package.json:19]: Sounds like the sensible way to do it.

**Comment by danielstevenberger** [comment]
I believe everything is resolved at this point. If there are no more issues I can finalize the README

**Comment by georgejolema** [comment]
hey @danielstevenberger I did some little adjustments to make possible use translations correctly and use the app with docker. From my side, everything looks good. Go ahead and work with README

**Review by georgejolema — APPROVED** [review]

File: README.md
Comment (tcorral) [inline README.md:3]: Are you using Jest for testing?

File: apps/golden-sample-app/src/app/ach-positive-pay/ach-positive-pay-journey-bundle.module.ts
Comment (tcorral) [inline apps/golden-sample-app/src/app/ach-positive-pay/ach-positive-pay-journey-bundle.module.ts:2]: Is this a local import?

File: apps/golden-sample-app/src/styles.scss
Comment (tcorral) [inline apps/golden-sample-app/src/styles.scss:3]: why this change?

File: docker-compose.yml
Comment (tcorral) [inline docker-compose.yml:12]: Will this not affect the problems with AWS?

File: docker-compose.yml
Comment (tcorral) [inline docker-compose.yml:16]: Is it save remove the realms of this url?

File: libs/ach-positive-pay/.eslintrc.json
Comment (tcorral) [inline libs/ach-positive-pay/.eslintrc.json:1]: Is this file a copy of transfer-journey, .eslintrc.json? It seems it has been renamed. Can you confirm both does exist?

File: libs/transactions/src/lib/components/text-filter/text-filter.component.spec.ts
Comment (tcorral) [inline libs/transactions/src/lib/components/text-filter/text-filter.component.spec.ts:1]: I miss some test about what this components does filter...

File: libs/transactions/src/lib/components/transaction-item/transaction-item.component.ts
Comment (tcorral) [inline libs/transactions/src/lib/components/transaction-item/transaction-item.component.ts:21]: You don't need to use square brackets when the name of the property of the object is just a word. Please use the dot notation.

File: libs/transfer/src/lib/components/make-transfer-summary/make-transfer-summary.component.spec.ts
Comment (tcorral) [inline libs/transfer/src/lib/components/make-transfer-summary/make-transfer-summary.component.spec.ts:1]: I miss a test checking that it shows a summary when a transfer is done. Can you add it, please?

File: libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts
Comment (tcorral) [inline libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts:5]: Why using `any` ?

File: libs/transfer/src/lib/views/make-transfer-success-view/make-transfer-success-view.component.ts
Comment (tcorral) [inline libs/transfer/src/lib/views/make-transfer-success-view/make-transfer-success-view.component.ts:9]: Dot notation, please.

File: libs/transfer/src/lib/views/make-transfer-summary-view/make-transfer-summary-view.component.ts
Comment (tcorral) [inline libs/transfer/src/lib/views/make-transfer-summary-view/make-transfer-summary-view.component.ts:10]: Dot notation, please.

File: libs/transfer/src/lib/views/make-transfer-view/make-transfer-view.component.ts
Comment (tcorral) [inline libs/transfer/src/lib/views/make-transfer-view/make-transfer-view.component.ts:15]: Dot notation, please.

**Review by tcorral — CHANGES_REQUESTED** [review]
Please, @danielstevenberger check my comments.

File: docker-compose.yml
Comment (georgejolema) [reply docker-compose.yml:12]: the problem with sandbox is fixed, and this is only used for local test

File: docker-compose.yml
Comment (georgejolema) [reply docker-compose.yml:16]: the rest of the url is added in the app. At the end you only provide api auth url.

File: README.md
Comment (danielstevenberger) [reply README.md:3]: Jest was setup for Unit Testing as discussed in the previous comments. I confirmed there were no issues in the Angular 13 branch.

File: apps/golden-sample-app/src/app/ach-positive-pay/ach-positive-pay-journey-bundle.module.ts
Comment (danielstevenberger) [reply apps/golden-sample-app/src/app/ach-positive-pay/ach-positive-pay-journey-bundle.module.ts:2]: This is how imports are used from libs to apps in Nx.

File: libs/transfer/src/lib/views/make-transfer-summary-view/make-transfer-summary-view.component.ts
Comment (danielstevenberger) [reply libs/transfer/src/lib/views/make-transfer-summary-view/make-transfer-summary-view.component.ts:10]: I changed because property 'title' comes from an index signature, so it must be accessed with ['title'].ts(4111)

I can disable noPropertyAccessFromIndexSignature if needed.

File: libs/transfer/src/lib/views/make-transfer-success-view/make-transfer-success-view.component.ts
Comment (danielstevenberger) [reply libs/transfer/src/lib/views/make-transfer-success-view/make-transfer-success-view.component.ts:9]: I changed because property 'title' comes from an index signature, so it must be accessed with ['title'].ts(4111)

I can disable noPropertyAccessFromIndexSignature if needed.

File: libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts
Comment (danielstevenberger) [reply libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts:5]: Do you know of a typing solution that would fix this. In Jest with Angular, I read to create the component as below. From my experience Jest is more developer friendly and significantly faster, however this is no createSpyObj function. Without createSpyObj, I wasn't sure how to get the typing correct for for creating the component. I tried combinations of partials, pick, and add.

 ```
   component = new TransactionsViewComponent(
      mockActivatedRoute,
      mockTransactionViewService,
      mockTransactionsCommunicationService
    );
```

File: libs/transfer/src/lib/components/make-transfer-summary/make-transfer-summary.component.spec.ts
Comment (danielstevenberger) [reply libs/transfer/src/lib/components/make-transfer-summary/make-transfer-summary.component.spec.ts:1]: I changed because property 'title' comes from an index signature, so it must be accessed with ['title'].ts(4111)

I can disable noPropertyAccessFromIndexSignature if needed.

File: libs/transactions/src/lib/components/transaction-item/transaction-item.component.ts
Comment (danielstevenberger) [reply libs/transactions/src/lib/components/transaction-item/transaction-item.component.ts:21]: I changed because property 'transactions' comes from an index signature, so it must be accessed with ['transactions'].ts(4111)

I can disable noPropertyAccessFromIndexSignature if needed.

File: libs/transactions/src/lib/components/text-filter/text-filter.component.spec.ts
Comment (danielstevenberger) [reply libs/transactions/src/lib/components/text-filter/text-filter.component.spec.ts:1]: I can add DOM manipulation testing if needed. Angular Test Bed with Karma/Jasmine may be better for this approach.

I made the change because PlayWright would test the the interaction of filtering and there is no functionality in this TypeScript file to test.

File: libs/ach-positive-pay/.eslintrc.json
Comment (danielstevenberger) [reply libs/ach-positive-pay/.eslintrc.json:1]: I'm not sure I understand your question, but all of the linting is working correctly.

File: apps/golden-sample-app/src/styles.scss
Comment (danielstevenberger) [reply apps/golden-sample-app/src/styles.scss:3]: VSCode is unable to find it's location with the previous import.

**Comment by danielstevenberger** [comment]
@tcorral I've replied to your comments, please let me know how you would like me to proceed.

File: libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts
Comment (danielstevenberger) [reply libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts:5]: I had watched a video testing this way using Jest in Angular, if there is a way to use a type I agree it would be better.

File: libs/transfer/src/lib/components/make-transfer-summary/make-transfer-summary.component.spec.ts
Comment (tcorral) [reply libs/transfer/src/lib/components/make-transfer-summary/make-transfer-summary.component.spec.ts:1]: ok

File: libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts
Comment (tcorral) [reply libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts:5]: Before your changes the previous test was using an interface. Check if it can still be used in yours.

File: libs/transfer/src/lib/views/make-transfer-summary-view/make-transfer-summary-view.component.ts
Comment (tcorral) [reply libs/transfer/src/lib/views/make-transfer-summary-view/make-transfer-summary-view.component.ts:10]: It's ok.

File: libs/transfer/src/lib/views/make-transfer-view/make-transfer-view.component.ts
Comment (tcorral) [reply libs/transfer/src/lib/views/make-transfer-view/make-transfer-view.component.ts:15]: It's ok.

File: libs/transfer/src/lib/views/make-transfer-success-view/make-transfer-success-view.component.ts
Comment (tcorral) [reply libs/transfer/src/lib/views/make-transfer-success-view/make-transfer-success-view.component.ts:9]: It's ok.

File: libs/transactions/src/lib/components/text-filter/text-filter.component.spec.ts
Comment (tcorral) [reply libs/transactions/src/lib/components/text-filter/text-filter.component.spec.ts:1]: How will it test this filtering.

File: libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts
Comment (danielstevenberger) [reply libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts:5]: I found a way to use better typing, I will change them all today.

File: libs/transactions/src/lib/components/text-filter/text-filter.component.spec.ts
Comment (danielstevenberger) [reply libs/transactions/src/lib/components/text-filter/text-filter.component.spec.ts:1]: The filter pipe will test the logic of the filtering. During an e2e test using Playwright, the filtering of the DOM would be tested. Jest without Angular Test bed focuses purely on the logic inside of each individual component which greatly improves speed of running Unit Tests.

File: libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts
Comment (danielstevenberger) [reply libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts:5]: Something like this could work. If you like this approach better I can update them all
```
  const snapshot: Partial<ActivatedRouteSnapshot> = {
    data: {
      title: 'someTitle',
    },
  };
  let scheduler: TestScheduler;
  let component: TransactionsViewComponent;
  const mockActivatedRoute = {
    snapshot: snapshot as ActivatedRouteSnapshot,
  };
  const mockTransactionsHttpService: Partial<TransactionsHttpService> = {
    transactions$: of(transactionsMock),
  };
  let mockTransactionsCommunicationService:
    | Partial<TransactionsCommunicationService>
    | undefined = {
    latestTransaction$: of(debitMockTransaction),
  };
  const createComponent = () => {
    component = new TransactionsViewComponent(
      mockActivatedRoute as ActivatedRoute,
      mockTransactionsHttpService as TransactionsHttpService,
      mockTransactionsCommunicationService as TransactionsCommunicationService
    );
  };
  beforeEach(() => {
    createComponent();
  });
```

Comment (jordydejong) [reply libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts:5]: Admittedly I have also used `any` for these cases in the past sometimes, however I think adding type safety with `Partial` is preferable. So changing a method in a service or component breaks test in compile time, instead of run time.

I'm curious about other opinions about removing Testbed. Personally I prefer testing without, because of performance, and it feels you are testing closer to "reality", for me personally. However Testbed seems to be the standard way of testing Angular components AFAIK, so not sure if what example we want to give others.

Comment (danielstevenberger) [reply libs/transfer/src/lib/services/make-transfer-permissions.service.spec.ts:5]: @jordydejong I will make these changes today.

**Comment by danielstevenberger** [comment]
I've added typing to the unit tests. The only thing left is to confirm the decision for no TestBed.

File: apps/golden-sample-app/src/app/app.component.spec.ts
Comment (jordydejong) [inline apps/golden-sample-app/src/app/app.component.spec.ts:6]: I think this typing is a step in the right direction. However if I (for instance) want to write this test, it leads to an error

```  it('should do foo for invalid access token', () => {
    mockOAuthService.hasValidAccessToken.mockReturnValueOnce(false);
    
    // expect...
  });
```
solution could be to use `const mockOAuthService: jest.Mocked<Pick<OAuthService, 'logOut' | 'hasValidAccessToken'>>`

but again, curious about other opinions on this

Comment (jordydejong) [reply apps/golden-sample-app/src/app/app.component.spec.ts:6]: In this case you would either have to do something like this
```
component = new AppComponent(
      mockOAuthService as unknown as OAuthService,
      mockLayoutService
    );
```

or use Testbed of course

Comment (santam85) [reply apps/golden-sample-app/src/app/app.component.spec.ts:6]: It's testing code, I'm not too concerned in using any for mocks and spies. What I do normally is using actual implementations and stub only the methods I need. 

For what concerns TestBed, removing it forces you to use full-on mocks for all dependencies, so sometimes it might be a bit easier to just rely on that and actual stubbed/spied real classes.

Comment (jordydejong) [reply apps/golden-sample-app/src/app/app.component.spec.ts:6]: Okay so while it is not necessarily my opinion :) the conclusion seems to be:

- using `any` for mocks is fine
- we use Testbed

correct @santam85 ?

Comment (danielstevenberger) [reply apps/golden-sample-app/src/app/app.component.spec.ts:6]: It depends on the constructor of the dependency being used to determine whether mocks or a new instance should be made. If playwright will be handling the e2e, focusing purely on the logic in the components is more effective. 

@jordydejong If you check locale-selector.component.spec.ts you can see how it's possible to test an invalid token in your example. This approach works works and is needed in cases anyways for any values set on initialization. 
That being said, I see the value of jest.Mocked now and will add it if we stick with this approach.

@santam85 
If the final decision is to use TestBed,  I will revert all the changes to the previous unit tests and setup Jasmine/Karma in Nx. If we use TestBed having createSpyObj is an important piece.

**Review by santam85 — APPROVED** [review]

**Review by tcorral — APPROVED** [review]

**Pull request merged**
